### PR TITLE
Support for pylint 3.x

### PR DESCRIPTION
**Pull request by matejsp**
Fixing https://github.com/pylint-dev/pylint-django/issues/405

### CODE REVIEW

**Review by Pierre-Sassoulas — APPROVED** [review]
You went out of your way to support both pylint 2 and 3 :ok_hand:

**Comment by carlio** [comment]
> You went out of your way to support both pylint 2 and 3 👌

This is something I have been clinging onto for a long time : new dependencies break the library, but dropping support for old dependencies breaks so many peoples' CI pipelines because they've pinned older versions. It has to be a careful mix of both  or people will just drop pylint-django (and prospector has this philosophy too).

I'm reluctant to say "this only supports the latest and greatest mix of Python+Django+Pylint" because that rules out the majority of people and companies who don't have the time to upgrade their dependencies, and honestly they're the ones who need CI static analysis the most in my opinion, to prevent building up too much technical debt since they don't have the capacity to go back and deal with it later.

I'm probably biased because a lot of my work is in helping people deal with technical debt caused by prioritizing new features and growth over maintenance.

Anyway that's my answer to your question on #405 @matejsp : "I didn't prepare PR because I don't know if you want to preserve backwards compatibility or just support pylint 3.x"

Sorry for being slow about this but I don't have much time right now, I'll get to it!

**Comment by Pierre-Sassoulas** [comment]
> I'm reluctant to say "this only supports the latest and greatest mix of Python+Django+Pylint" even though that's easier for me, because that rules out the majority of people and companies who don't have the time to upgrade their dependencies, and honestly they're the ones who need CI static analysis the most in my opinion, to prevent building up too much technical debt since they don't have the capacity to go back and deal with it later.

Personally, I'm following the system, the language, or the framework's EOL guidelines (debian/python/django) witohut feeling guilty about it. If upgrading would be very costly for a company, then they can pay you a fraction of the upgrade cost so you'll maintain your code for longer than the EOL (never seen that done for a lib, but I guess Ubuntu can do it with financial success with their Expanded Security Maintenance). Everyone can use the existing old version from 4 year ago that is compatible with their old stack (and ignore the CVEs the same way they ignore the OS/language/framework CVEs).  For this reason it's even arguably not that problematic to support for less than the EOL as long as one version of your lib is compatible with interpreters that are still not EOL (i.e. pylint-django could drop support for python 3.8 / pylint 2.17.7 right now, there's still usable versions for python 3.8 / pylint 2.17.7). To be frank, if deploying prod code on an OS that is affected by heart-bleed in 2020+ is not enough of a reason for a company to stop prioritizing business above all else and finally upgrade, then I don't know why this company should survive on your back. There's no reason you in particular should suffer to allow them to survive without ripping any of the benefits.

That said it's nice to stay compatible when the cost is low or if (hypothetically) you're compensated for your work.

File: pylint_django/checkers/foreign_key_strings.py [hunk 1]
```diff
@@ -95,7 +92,12 @@ def open(self):
             # this means that Django wasn't able to configure itself using some defaults
             # provided (likely in a DJANGO_SETTINGS_MODULE environment variable)
             # so see if the user has specified a pylint option
-            if self.config.django_settings_module is None:
+            if hasattr(self, "linter"):
+                django_settings_module = self.linter.config.django_settings_module
```
Comment (twanwv) [inline pylint_django/checkers/foreign_key_strings.py:96]: This line is hit at my end (`self.config` is `None`) meaning that at line 113 the `self.config.django_settings_module` is resulting in an error, I think that 113 should be: `settings.configure(Settings(django_settings_module))` which worked for me.

**Review by Pierre-Sassoulas — COMMENTED** [review]
I've tested locally and I confirm it works for me.

``pylint --version``
```
pylint 3.0.1
astroid 3.0.0
Python 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0]
```

```
pip3 freeze|grep pylint-django
pylint-django @ git+https://github.com/matejsp/pylint-django.git@bab54286b0d6d5a06ffec28198a61d7ebf6a38b4
```

After fixing an [unrelated hack I did before based on a stackoverflow answer](https://stackoverflow.com/a/74242848/2519059) that "just worked":
```diff
-init-hook="import sys; import os; from pylint.config import find_pylintrc; sys.path.append(os.path.dirname(find_pylintrc()))"
+init-hook="import os;os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'toolkit.settings')"
```

**Comment by carlio** [comment]
I have been trying to get the CI working (see https://github.com/pylint-dev/pylint-django/pull/400#issuecomment-1773980016) but it's a bit of a pain so I'm going to merge this and make a bugfix release first to free up peoples' own CI builds.

**Comment by carlio** [comment]
Finally managed to make a bugfix release to include this. `2.5.4` is on PyPI. Thanks again @matejsp :-)

I'll get on to cleaning up the pylint-django CI process next.

**Pull request merged**